### DISCUSSION TITLE
Create matrix with multiple OS and env for each one

### DISCUSSION THREAD

**Discussion by patrickelectric** [community/community#25468] [Actions]
<p>I’m trying to create something that share the same commands, and each OS can configure the commands via env variables. Is such thing possible ? I’m trying with this approach but it’s failing:</p>
<pre><code>jobs:
  build:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        include:
          os: macos-latest
            env:
              - TARGET: x86_64-apple-darwin
              - COMPILER: clang
              - LINKER: clang

          os: ubuntu-latest
            env:
              - TARGET: armv7-unknown-linux-musleabihf
              - COMPILER: arm-linux-gnueabihf-gcc-5
              - LINKER: gcc-5-arm-linux-gnueabihf

          os: ubuntu-latest
            env:
              - TARGET: x86_64-unknown-linux-musl
              - COMPILER: gcc
              - LINKER: gcc
</code></pre>

**Comment by Yanjingzhu** [comment]
<p>I am very glad to see you here. According to your scenario, I would suggest you use including configurations in a matrix build.</p>
<p> You can set os as matrix and then use include keyword to specify TARGET, COMPILER and LINKER. In steps, you can use ${{ matrix.TARGET }} to get the value. Please check the example, it will create three jobs::</p>
<pre><code>name: test matrix
on: push
jobs:
  build:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [macos-latest, ubuntu-latest, ubuntu-latest]
        include:
        - os: macos-latest
          TARGET: x86_64-apple-darwin
          COMPILER: clang
          LINKER: clang

        - os: ubuntu-latest
          TARGET: armv7-unknown-linux-musleabihf
          COMPILER: arm-linux-gnueabihf-gcc-5
          LINKER: gcc-5-arm-linux-gnueabihf

        - os: ubuntu-latest
          TARGET: x86_64-unknown-linux-musl
          COMPILER: gcc
          LINKER: gcc
    steps:
     - uses: actions/checkout@v1
     - run: echo ${{matrix.TARGET}}
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/a8dfe86681d79bddee051fcd4bbbbec556d9e643.png" data-download-href="/uploads/short-url/o5VVZlHszPg0W1I5nMMpKCopmF5.png?dl=1" title="matrix result.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181091241-f9ac577e-fe0a-4fd2-b0ab-649454accf1f.png" alt="matrix result.png" data-base62-sha1="o5VVZlHszPg0W1I5nMMpKCopmF5" width="690" height="289" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/a8dfe86681d79bddee051fcd4bbbbec556d9e643_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">matrix result.png</span><span class="informations">751×315 16.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>For matrix syntax, please refer to this official document:  <a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fhelp.github.com%2Fen%2Factions%2Fautomating-your-workflow-with-github-actions%2Fworkflow-syntax-for-github-actions%23example-including-configurations-in-a-matrix-build&amp;data=02%7C01%7Cv-yjingz%40microsoft.com%7C6349a613f22c4e06d03908d76bf3101d%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637096563999012399&amp;sdata=F23dnDA0R7WxIOL15%2FMA7cQa6HWmxMHfcEtHOBxNe9s%3D&amp;reserved=0" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#example-including-configurations-in-a-matrix-build</a></p>

**Comment by geekswamp** [comment]
<p>Thank you very much <a class="mention" href="/yanjingzhu">@yanjingzhu</a> </p>

**Comment by sergeyklay** [comment]
<p>Just tested this configuration and have to say it does not work. Take a look at screenshots:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/4/5/45a900d46322ff3b4d77d4bea856dd7388a8b29e.png" data-download-href="/uploads/short-url/9Wf23KGMFV9efzx3CKZTKzX7ARw.png?dl=1" title="Screenshot 2020-10-09 162725" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/1256298/181091243-21993b58-5821-4c52-9569-240db0480512.png" alt="Screenshot 2020-10-09 162725" data-base62-sha1="9Wf23KGMFV9efzx3CKZTKzX7ARw" width="690" height="233" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/4/5/45a900d46322ff3b4d77d4bea856dd7388a8b29e_2_690x233.png, https://aws1.discourse-cdn.com/github/original/3X/4/5/45a900d46322ff3b4d77d4bea856dd7388a8b29e.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/4/5/45a900d46322ff3b4d77d4bea856dd7388a8b29e.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/4/5/45a900d46322ff3b4d77d4bea856dd7388a8b29e_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot 2020-10-09 162725</span><span class="informations">1009×342 20.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/a/c/aca712e6a78e0812325bbb9de6b3720a433fdc3d.png" data-download-href="/uploads/short-url/oDm4Tul8DgP3mXj9wEwwNKpy4K9.png?dl=1" title="Screenshot 2020-10-09 162706" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/1256298/181091244-5720f3b4-2d73-40a1-9dfd-e9b29dd777ee.png" alt="Screenshot 2020-10-09 162706" data-base62-sha1="oDm4Tul8DgP3mXj9wEwwNKpy4K9" width="690" height="216" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/a/c/aca712e6a78e0812325bbb9de6b3720a433fdc3d_2_690x216.png, https://aws1.discourse-cdn.com/github/optimized/3X/a/c/aca712e6a78e0812325bbb9de6b3720a433fdc3d_2_1035x324.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/a/c/aca712e6a78e0812325bbb9de6b3720a433fdc3d.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/a/c/aca712e6a78e0812325bbb9de6b3720a433fdc3d_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot 2020-10-09 162706</span><span class="informations">1111×349 20.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Both jobs are <code>x86_64-unknown-linux-musl</code></p>

**Comment by sergeyklay** [comment, marked as answer]
<p>Correct workflow is:</p>
<pre><code class="lang-auto">name: test matrix
on: push
jobs:
  build:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:

        include:
          - os: macos-latest
            TARGET: x86_64-apple-darwin
            COMPILER: clang
            LINKER: clang

          - os: ubuntu-latest
            TARGET: armv7-unknown-linux-musleabihf
            COMPILER: arm-linux-gnueabihf-gcc-5
            LINKER: gcc-5-arm-linux-gnueabihf

          - os: ubuntu-latest
            TARGET: x86_64-unknown-linux-musl
            COMPILER: gcc
            LINKER: gcc
    steps:
      - uses: actions/checkout@v1
      - run: echo ${{matrix.TARGET}}
</code></pre>
<p>Do not add <code>os</code> as proposed above</p>

**Comment by Simran-B** [comment]
<p>As far as I’ve seen in a little experiment on my own, using the same key multiple times under <code>include</code> (here: <code>ubuntu-latest</code>) overwrites the job configuration and the last entry wins. That is, the matrix is configured for 1x macOS and 2x Ubuntu, but first Ubuntu configuration is overwritten by the second, so both end up with <code>TARGET: x86_64-unknown-linux-musl</code> etc.</p>
<p><a class="mention" href="/sergeyklay">@sergeyklay</a>’s solution to use <code>include</code> only and no other keys under <code>matrix</code> is the best option IMO. Alternatively, one could use a unique key, but this solution doesn’t add anything over using <code>include</code> only:</p>
<pre><code class="lang-auto">build: [1, 2, 3]
include:
  - build: 1
    os: macos-latest
    ...

  - build: 2
    os: ubuntu-latest
    ...

  - build: 3
    os: ubuntu-latest
    ...
</code></pre>
